rfcs/0005-nix-encryption.md

```diff
@@ -0,0 +1,273 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* The main downside of this approach is that encrypted files need to
+  be decrypted before they can be used. Thus, for example, a daemon
+  configuration file cannot be used directly from the Nix store by
+  passing something like `--config /nix/store/.../foo.conf` to the
+  daemon. Instead, we have to decrypt the configuration file to some
+  suitably secure temporary location in a pre-start script, and then
+  pass that location (e.g. `--config /tmp/.../foo.conf`).
+
+  In NixOS, we can provide some helper functions to reduce boilerplate
+  code dealing with encrypted files in systemd services. For example,
+  we can add an option to allow modules to declare files that need
+  decryption:
+
+    ```
+    security.encryptedFiles = [ wpaSupplicantConf ];
+    ```
+
+  which would be decrypted by the activation script to a well-known
+  location such as `/var/secrets/<storepath>`. This way, services can
+  refer to the decrypted path easily:
+
+    ```
+    systemd.services.wpa_supplicant.serviceConfig.Exec =
+      "wpa_supplicant -c /var/secrets/${wpaSupplicantConf}";
+      # i.e. ... -c /var/secrets/nix/store/.../wpa_supplicant.conf
+    ```
+
+  The above can also be extended to support access by non-root users,
+  e.g.
+
+    ```
+    security.encryptedFiles = [
+      { file = httpdConf;
+        owner = "httpd";
+      }
+    ];
+    ```
+
+  An alternative is to write a little FUSE filesystem that
+  transparently decrypts files on the underlying filesystem. For
+  example, the filesystem mounted on /var/secrets (accessible only to
+  root) would automatically decrypt files from /nix/store. But that's
+  probably over-engineering it.
+
+* This approach only allows secrets to be stored at evaluation time
+  and used at runtime. Secrets cannot be created or used at build time
+  (except by copying encrypted secrets verbatim), since no keys are
+  available at build time.
+
+* The fact that encrypted secrets are world-readable allows attackers
+  to mount (offline) brute-force attacks. Depending of how we handle
+  nonces (see below), it may also be possible for attackers to observe
+  that secret values have changed, that identical secrets are used in
+  multiple places, or that secrets have changed back to a previous
+  value.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Access control
+
+The main alternative is to add some notion of access control
+(i.e. ownership and permissions) to Nix store paths. (See
+e.g. https://github.com/NixOS/nix/pull/329.) This has the advantage
+that secret files can be used "directly" by a sufficiently privileged
+user. However, adding access control raises many issues:
+
+* Nix has always assumed a world-readable store, so we'd have to
+  carefully audit that nothing in the daemon can leak the contents of
+  store paths.
+
+* Users should be allowed to export closures that they built, even
```

Comment:
Perhaps instead of checking the ACL's of the store paths, use the evaluator to determine whether the user can copy the closure?